rskj-core/src/main/java/org/ethereum/util/RLP.java

```diff
@@ -191,43 +199,62 @@ public static BigInteger decodeBigInteger(byte[] data, int index) {
         return BigIntegers.fromUnsignedByteArray(bytes);
     }
 
+    private static void assertLengthOfLength(byte[] payload,byte lengthOfLength) {
+        if (lengthOfLength >= payload.length) {
+            throw new RLPException("lengthOfLength cannot be higher than payload length");
+        }
+    }
+
     public static int getNextElementIndex(byte[] payload, int pos) {
 
         if (pos >= payload.length) {
             return -1;
         }
 
+        int nextElementIndex = -1;
+
         if ((payload[pos] & 0xFF) >= OFFSET_LONG_LIST) {
             byte lengthOfLength = (byte) (payload[pos] - OFFSET_LONG_LIST);
+
+            assertLengthOfLength(payload, lengthOfLength);
+
             int length = calcLength(lengthOfLength, payload, pos);
-            return pos + lengthOfLength + length + 1;
+            nextElementIndex = pos + lengthOfLength + length + 1;
         }
         if ((payload[pos] & 0xFF) >= OFFSET_SHORT_LIST
                 && (payload[pos] & 0xFF) < OFFSET_LONG_LIST) {
 
             byte length = (byte) ((payload[pos] & 0xFF) - OFFSET_SHORT_LIST);
-            return pos + 1 + length;
+            nextElementIndex = pos + 1 + length;
         }
         if ((payload[pos] & 0xFF) >= OFFSET_LONG_ITEM
                 && (payload[pos] & 0xFF) < OFFSET_SHORT_LIST) {
 
             byte lengthOfLength = (byte) (payload[pos] - OFFSET_LONG_ITEM);
+
+            assertLengthOfLength(payload, lengthOfLength);
+
             int length = calcLength(lengthOfLength, payload, pos);
-            return pos + lengthOfLength + length + 1;
+            nextElementIndex = pos + lengthOfLength + length + 1;
         }
         if ((payload[pos] & 0xFF) > OFFSET_SHORT_ITEM
                 && (payload[pos] & 0xFF) < OFFSET_LONG_ITEM) {
 
             byte length = (byte) ((payload[pos] & 0xFF) - OFFSET_SHORT_ITEM);
-            return pos + 1 + length;
+            nextElementIndex = pos + 1 + length;
         }
         if ((payload[pos] & 0xFF) == OFFSET_SHORT_ITEM) {
-            return pos + 1;
+            nextElementIndex = pos + 1;
         }
         if ((payload[pos] & 0xFF) < OFFSET_SHORT_ITEM) {
-            return pos + 1;
+            nextElementIndex = pos + 1;
         }
-        return -1;
+
+        if (pos + 1 <= nextElementIndex && nextElementIndex < payload.length) {
+            return nextElementIndex;
+        }
+
+        throw new RLPException("Next element index out of next pos and payload length range");
```

Comment:
so `-1` isn't legitimate value anymore? then shouldn't be thrown an exception in [here](https://github.com/rsksmart/rskj/pull/3052/files#diff-12b253aa5c996cabe09f5ab95a1f6a4123ed6faf5fc56518fcf549644cb0d47bR211) too?

Reply:
There are other conditions where the `-1` will be returned, according to the context, we just want to cover specific 2 conditions, you can see nextElementIndex is itinitialized as `-1` because of that

Reply:
there's only one usage of this method in FrameCodec. As I can see, you already replaced one return of -1 with throwing the RLPException error. Wouldn't it make sense to replace the other one?

Reply:
Addressed